options/customizable.cc

```diff
@@ -0,0 +1,139 @@
+
+#include "rocksdb/convenience.h"
+#include "rocksdb/status.h"
+#include "util/string_util.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+std::string Customizable::GetOptionName(const std::string& long_name) const {
+  const std::string& name = Name();
+  size_t name_len = name.size();
+  if (long_name.size() > name_len + 1 &&
+      long_name.compare(0, name_len, name) == 0 &&
+      long_name.at(name_len) == '.') {
+    return long_name.substr(name_len + 1);
+  } else {
+    return Configurable::GetOptionName(long_name);
+  }
+}
+
+#ifndef ROCKSDB_LITE
+Status Customizable::GetOption(const ConfigOptions& config_options,
+                               const std::string& opt_name,
+                               std::string* value) const {
+  if (opt_name == kIdPropName) {
+    *value = GetId();
+    return Status::OK();
+  } else {
+    return Configurable::GetOption(config_options, opt_name, value);
+  }
+}
+
+std::string Customizable::SerializeOptions(const ConfigOptions& config_options,
+                                           const std::string& prefix) const {
+  std::string result;
+  std::string parent;
+  if (!config_options.IsShallow()) {
+    parent = Configurable::SerializeOptions(config_options, "");
+  }
+  if (parent.empty()) {
+    result = GetId();
+  } else {
+    result.append(prefix + kIdPropName + "=" + GetId() +
+                  config_options.delimiter);
+    result.append(parent);
+  }
+  return result;
+}
+
+#endif  // ROCKSDB_LITE
+
+bool Customizable::AreEquivalent(const ConfigOptions& config_options,
+                                 const Configurable* other,
+                                 std::string* name) const {
+  if (config_options.sanity_level > ConfigOptions::kSanityLevelNone &&
+      this != other) {
+    const Customizable* custom = reinterpret_cast<const Customizable*>(other);
+    if (GetId() != custom->GetId()) {
+      *name = kIdPropName;
+      return false;
+    } else if (config_options.sanity_level >
+               ConfigOptions::kSanityLevelLooselyCompatible) {
+      bool matches = Configurable::AreEquivalent(config_options, other, name);
+      return matches;
+    }
+  }
+  return true;
+}
+
+Status Customizable::ConfigureNewObject(
+    const ConfigOptions& config_options_in, Customizable* object,
+    const std::string& id, const std::string& base_opts,
+    const std::unordered_map<std::string, std::string>& opts) {
+  if (object != nullptr) {
+    ConfigOptions config_options = config_options_in;
+    config_options.invoke_prepare_options = false;
+    if (!base_opts.empty()) {
+#ifndef ROCKSDB_LITE
+      // Don't run prepare options on the base, as we would do that on the
+      // overlay opts instead
+      Status status = object->ConfigureFromString(config_options, base_opts);
+      if (!status.ok()) {
+        return status;
+      }
+#endif  // ROCKSDB_LITE
+    }
+    if (!opts.empty()) {
+      return object->ConfigureFromMap(config_options, opts);
```

Comment:
So here, if base_opts and opts are not empty, we configure the object twice, right? Some of the options in base_opts will be overwritten by opts?

Reply:
Yes, we would configure the object twice, but not necessarily the same parts